src/Shared/PackageManagers/ProjectManagerFactory.cs

```diff
@@ -8,55 +8,153 @@ namespace Microsoft.CST.OpenSource.PackageManagers
     using System.Linq;
     using System.Net.Http;
 
-    public static class ProjectManagerFactory
+    public class ProjectManagerFactory
     {
         /// <summary>
-        /// Get an appropriate project manager for package given its PackageURL.
+        /// The delegate to use when defining the function to create a project manager with.
+        /// The only runtime parameter we need is the destination directory. Everything else can be defined in the constructor.
         /// </summary>
-        /// <param name="purl">The <see cref="PackageURL"/> for the package to create the project manager for.</param>
-        /// <param name="httpClientFactory"> The <see cref="IHttpClientFactory"/> for the project manager to use for making Http Clients to make web requests.</param>
-        /// <param name="destinationDirectory">The directory to use to store any downloaded packages.</param>
-        /// <returns> BaseProjectManager object </returns>
-        public static BaseProjectManager? CreateProjectManager(PackageURL purl, IHttpClientFactory? httpClientFactory = null, string? destinationDirectory = null)
+        public delegate BaseProjectManager? ConstructProjectManager(string destinationDirectory = ".");
+
+        /// <summary>
+        /// The dictionary of project managers.
+        /// </summary>
+        private readonly Dictionary<string, ConstructProjectManager> _projectManagers;
+
+        /// <summary>
+        /// Initializes a new instance of a <see cref="ProjectManagerFactory"/>.
+        /// </summary>
+        /// <param name="overrideManagers">Use a custom set of managers to override the defaults with.</param>
+        /// <param name="httpClientFactory">The <see cref="IHttpClientFactory"/> to use in the project managers.</param>
+        public ProjectManagerFactory(Dictionary<string, ConstructProjectManager> overrideManagers, IHttpClientFactory? httpClientFactory = null) : this(httpClientFactory)
         {
-            if (projectManagers.Count == 0)
-            {
-                projectManagers.AddRange(typeof(BaseProjectManager).Assembly.GetTypes()
-               .Where(type => type.IsSubclassOf(typeof(BaseProjectManager))));
-            }
-            // Use reflection to find the correct package management class
-            Type? downloaderClass = projectManagers
-               .Where(type => type.Name.Equals($"{purl.Type}ProjectManager",
-                                               StringComparison.InvariantCultureIgnoreCase))
-               .FirstOrDefault();
-
-            if (downloaderClass != null)
+            overrideManagers.ToList().ForEach(pair => SetManager(pair.Key, pair.Value));
```

Comment:
This seems like an unnecessary allocation.

```suggestion
            foreach(var pair in overrideManagers)
            {
                SetManager(pair.Key, pair.Value);
            }
```